src/Cardano/Wallet/Network/HttpBridge.hs

```diff
@@ -132,79 +132,72 @@ fetchBlocksFromTip network start tipHash =
+        :: Word64 -> ExceptT ErrNetworkUnreachable m [Block]
     , getNetworkTip
-        :: ExceptT e m (Hash "BlockHeader", BlockHeader)
+        :: ExceptT ErrNetworkUnreachable m (Hash "BlockHeader", BlockHeader)
     , postSignedTx
-        :: SignedTx -> ExceptT e m ()
+        :: SignedTx -> ExceptT ErrPostTx m ()
     }
 
 -- | Construct a new network layer
 mkHttpBridge
-    :: Manager -> BaseUrl -> NetworkName -> HttpBridge IO HttpBridgeError
+    :: Manager -> BaseUrl -> NetworkName -> HttpBridge IO
 mkHttpBridge mgr baseUrl network = HttpBridge
-    { getBlock = \hash -> do
+    { getBlock = \hash -> ExceptT $ do
         hash' <- hashToApi' hash
-        run (getApiT <$> cGetBlock network hash')
-    , getEpoch = \ep ->
-        run (map getApiT <$> cGetEpoch network (EpochIndex ep))
-    , getNetworkTip =
-        run (blockHeaderHash <$> cGetNetworkTip network)
-    , postSignedTx =
-        -- TODO: We would be good to parse errors from the node
-        -- like "Failed to send to peers: Blockchain protocol error"
-        -- and "Transaction failed verification: transaction has no inputs"
-        void . run . (cPostSignedTx network) . ApiT
+        run (getApiT <$> cGetBlock network hash') >>= defaultHandler
+
+    , getEpoch = \ep -> ExceptT $ do
+        run (map getApiT <$> cGetEpoch network (EpochIndex ep)) >>= \case
+            Left (FailureResponse e) | responseStatusCode e == status404 ->
+                return $ Right []
+            x -> defaultHandler x
+
+    , getNetworkTip = ExceptT $ do
+        run (blockHeaderHash <$> cGetNetworkTip network) >>= defaultHandler
+
+    , postSignedTx = \tx -> void $ ExceptT $ do
+        let e0 = "Failed to send to peers: Blockchain protocol error"
+        run (cPostSignedTx network (ApiT tx)) >>= \case
+            Left (FailureResponse e)
+                | responseStatusCode e == status400 -> do
+                    let msg = T.decodeUtf8 $ BL.toStrict $ responseBody e
+                    return $ Left $ ErrPostTxBadRequest msg
+            Left (FailureResponse e)
+                | responseStatusCode e == status500 && responseBody e == e0 -> do
+                    let msg = T.decodeUtf8 $ BL.toStrict $ responseBody e
+                    return $ Left $ ErrPostTxProtocolFailure msg
+            x -> left ErrPostTxNetworkUnreachable <$> defaultHandler x
     }
   where
-    run :: ClientM a -> ExceptT HttpBridgeError IO a
-    run query = ExceptT $ (first convertError) <$> runClientM query env
-
-    env = mkClientEnv mgr baseUrl
+    run :: ClientM a -> IO (Either ServantError a)
+    run query = runClientM query (mkClientEnv mgr baseUrl)
+
+    defaultHandler
+        :: Either ServantError a
+        -> IO (Either ErrNetworkUnreachable a)
+    defaultHandler = \case
+        Right c -> return $ Right c
+        Left (ConnectionError e) ->
+            return $ Left $ ErrNetworkUnreachable e
+        Left e ->
+            throwM e
```

Comment:
The exception type parameter was not a good idea, so it's now better with that removed.

But this handler needs more cases - `FailureResponse` at least. This change has broken the benchmarks because the network tip endpoint can return 404. The way to recover from this exception is to wait and retry.